code/go/0chain.net/smartcontract/storagesc/readpool.go

```diff
@@ -368,41 +282,34 @@ func (ssc *StorageSmartContract) readPoolUnlock(t *transaction.Transaction,
 		req.PoolOwner = t.ClientID
 	}
 
-	isFunded, err := ssc.isFundedPool(t.ClientID, req.PoolID, balances)
-	if err != nil {
-		return "", common.NewError("read_pool_unlock_failed", err.Error())
-	}
-	if !isFunded {
-		return "", common.NewErrorf("read_pool_unlock_failed",
-			"%s did not fund pool %s", t.ClientID, req.PoolID)
-	}
-
-	var rp *readPool
-	if rp, err = ssc.getReadPool(req.PoolOwner, balances); err != nil {
-		return "", common.NewError("read_pool_unlock_failed", err.Error())
+	txValue := -(state.Balance(t.Value)) // subtract from read_pool balance
+	if err = ssc.modifyReadPoolBalance(t.ClientID, req.IsOwner, txValue, balances); err != nil {
+		return "", err
 	}
 
-	var ap *allocationPool
-	if ap, err = rp.take(req.PoolID, t.CreationDate); err != nil {
-		return "", common.NewError("read_pool_unlock_failed", err.Error())
-	}
+	return
+}
 
-	transfer, resp, err = ap.EmptyPool(ssc.ID, t.ClientID,
-		common.ToTime(t.CreationDate))
-	if err != nil {
-		return "", common.NewError("read_pool_unlock_failed", err.Error())
+func (ssc *StorageSmartContract) modifyReadPoolBalance(clientID string, isOwner bool, txValue state.Balance, balances cstate.StateContextI) error {
+	var (
+		rp  *newReadPool
+		err error
+	)
+	mutex.Lock()
+	defer mutex.Unlock()
+	if rp, err = ssc.getNewReadPool(clientID, balances); err != nil {
+		return common.NewError("read_pool_unlock_failed", err.Error())
 	}
 
-	if err = balances.AddTransfer(transfer); err != nil {
-		return "", common.NewError("read_pool_unlock_failed", err.Error())
-	}
+	//adjust balance
```

Comment:
We need to put back the balances.AddTransfer call, otherwise no tokens will actually be loked.

Reply:
We need to talk about the implementation of AddTransfer in this context.